jaeger-core/src/main/java/com/uber/jaeger/propagation/TextMapCodec.java

```diff
@@ -61,39 +71,102 @@ public void inject(SpanContext spanContext, TextMap carrier) {
     }
   }
 
+  /**
+   * Populates SpanContext based on the header values available in the carrier.
+   *
+   * Rules followed are:
+   * 1. When trace ID is present
+   *    - baggage from uberctx- is respected
+   *    - baggage from jaeger-baggage is not respected
+   * 2. When trace ID is not present
+   *    - baggage from uberctx- is not respected
+   *    - baggage from jaeger-baggage is respected
+   * 3. When neither trace ID, nor debug ID is present, null is returned.
+   * 4. Debug ID is always respected regardless of other data
+   *
+   * @param carrier The iterable to parse the headers out of
+   * @return SpanContext populated with information from headers, based on aforementioned rules.
+   */
   @Override
   public SpanContext extract(TextMap carrier) {
-    SpanContext context = null;
     Map<String, String> baggage = null;
-    String debugId = null;
+    Map<String, String> debugBaggage = null;
+    @Nullable String debugId = null;
+    @Nullable String traceId = null;
     for (Map.Entry<String, String> entry : carrier) {
       // TODO there should be no lower-case here
       String key = entry.getKey().toLowerCase(Locale.ROOT);
       if (key.equals(contextKey)) {
-        context = SpanContext.contextFromString(decodedValue(entry.getValue()));
+        traceId = decodedValue(entry.getValue());
       } else if (key.equals(Constants.DEBUG_ID_HEADER_KEY)) {
         debugId = decodedValue(entry.getValue());
       } else if (key.startsWith(baggagePrefix)) {
         if (baggage == null) {
           baggage = new HashMap<String, String>();
         }
         baggage.put(keys.unprefixedKey(key, baggagePrefix), decodedValue(entry.getValue()));
+      } else if (key.equals(MANUAL_BAGGAGE_HEADER_KEY)) {
+        String decodedDebugBaggageItems = decodedValue(entry.getValue());
+        debugBaggage = getDebugHeadersAsBaggage(decodedDebugBaggageItems);
```

Comment:
note that the header values separated by comma can be presented as duplicate headers. I.e. if sender sets header `X: a=b, x=y`, the server might receive

```
X: a=b
X: x=y
```
So for correctness this "debugBaggage" needs to be cumulative. Also, manualBaggage would be a better name, given the constant name.

Reply:
Ooh - good point. Gonna make both changes